cmd/snap-confine/seccomp-support.c

```diff
@@ -17,869 +17,140 @@
-#include <sys/utsname.h>
-#include <termios.h>
 #include <unistd.h>
 
-// The XFS interface requires a 64 bit file system interface
-// but we don't want to leak this anywhere else if not globally
-// defined.
-#ifndef _FILE_OFFSET_BITS
-#define _FILE_OFFSET_BITS 64
-#include <xfs/xqm.h>
-#undef _FILE_OFFSET_BITS
-#else
-#include <xfs/xqm.h>
-#endif
-
-#include <seccomp.h>
+#include <linux/filter.h>
+#include <linux/seccomp.h>
 
 #include "../libsnap-confine-private/secure-getenv.h"
 #include "../libsnap-confine-private/string-utils.h"
 #include "../libsnap-confine-private/utils.h"
 
-#define sc_map_add(X) sc_map_add_kvp(#X, X)
-
-// libseccomp maximum per ARG_COUNT_MAX in src/arch.h
-#define SC_ARGS_MAXLENGTH	6
-#define SC_MAX_LINE_LENGTH	82	// 80 + '\n' + '\0'
-
-enum parse_ret {
-	PARSE_INVALID_SYSCALL = -2,
-	PARSE_ERROR = -1,
-	PARSE_OK = 0,
-};
-
-struct preprocess {
-	bool unrestricted;
-	bool complain;
-};
-
-/*
- * arg_cmp contains items of type scmp_arg_cmp (from SCMP_CMP macro) and
- * length is the number of items in arg_cmp that are active such that if
- * length is '3' arg_cmp[0], arg_cmp[1] and arg_cmp[2] are used, when length
- * is '1' only arg_cmp[0] and when length is '0', none are used.
- */
-struct seccomp_args {
-	int syscall_nr;
-	unsigned int length;
-	struct scmp_arg_cmp arg_cmp[SC_ARGS_MAXLENGTH];
-};
-
-struct sc_map_entry {
-	ENTRY *e;
-	ENTRY *ep;
-	struct sc_map_entry *next;
-};
-
-struct sc_map_list {
-	struct sc_map_entry *list;
-	int count;
-};
-
 static char *filter_profile_dir = "/var/lib/snapd/seccomp/profiles/";
-static struct hsearch_data sc_map_htab;
-static struct sc_map_list sc_map_entries;
-
-/*
- * Setup an hsearch map to map strings in the policy (eg, AF_UNIX) to
- * scmp_datum_t values. Abstract away hsearch implementation behind sc_map_*
- * functions in case we want to swap this out.
- *
- * sc_map_init()		- initialize the hash map via linked list of
- * 				  of entries
- * sc_map_add_kvp(key, value)	- create entry from key/value pair and add to
- * 				  linked list
- * sc_map_search(s)	- if found, return scmp_datum_t for key, else set errno
- * sc_map_destroy()	- destroy the hash map and linked list
- */
-static scmp_datum_t sc_map_search(const char *s)
-{
-	ENTRY e = { 0, 0 };
-	ENTRY *ep = NULL;
-	scmp_datum_t val = 0;
-	errno = 0;
-
-	e.key = (char *)s;
-	if (hsearch_r(e, FIND, &ep, &sc_map_htab) == 0)
-		die("hsearch_r failed for %s", s);
-
-	if (ep != NULL) {
-		scmp_datum_t *val_p = NULL;
-		val_p = ep->data;
-		val = *val_p;
-	} else
-		errno = EINVAL;
 
-	return val;
-}
-
-static void sc_map_add_kvp(const char *key, scmp_datum_t value)
-{
-	struct sc_map_entry *node;
-	scmp_datum_t *value_copy;
-
-	node = malloc(sizeof(*node));
-	if (node == NULL)
-		die("Out of memory creating sc_map_entries");
-
-	node->e = malloc(sizeof(*node->e));
-	if (node->e == NULL)
-		die("Out of memory creating ENTRY");
+// MAX_BPF_SIZE is an arbitrary limit.
+const int MAX_BPF_SIZE = 640 * 1024;
```

Comment:
I'd rather keep in on the smallishly reasonable side cause if things all of a sudden get big, that would indicate something we'd want to look into. Note, I only tested this on amd64.